### PR TITLE
[SPARK-22266][SQL] The same aggregate function was evaluated multiple times

### PR DESCRIPTION
**Pull request by maryannxue**
## What changes were proposed in this pull request?

To let the same aggregate function that appear multiple times in an Aggregate be evaluated only once, we need to deduplicate the aggregate expressions. The original code was trying to use a "distinct" call to get a set of aggregate expressions, but did not work, since the "distinct" did not compare semantic equality. And even if it did, further work should be done in result expression rewriting.
In this PR, I changed the "set" to a map mapping the semantic identity of a aggregate expression to itself. Thus, later on, when rewriting result expressions (i.e., output expressions), the aggregate expression reference can be fixed.

## How was this patch tested?

Added a new test in SQLQuerySuite

### CODE REVIEW

**Comment by gatorsmile** [comment]
Update the title to `[SPARK-22266][SQL] The same aggregate function was evaluated multiple times`

**Comment by gatorsmile** [comment]
add to whitelist

**Comment by cloud-fan** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala [hunk 1]
```diff
@@ -205,14 +205,15 @@ object PhysicalAggregation {
...
       // In order to avoid evaluating an individual aggregate function multiple times, we'll
-      // build a set of the distinct aggregate expressions and build a function which can
+      // build a map of the distinct aggregate expressions and build a function which can
       // be used to re-write expressions so that they reference the single copy of the
       // aggregate function which actually gets computed.
-      val aggregateExpressions = resultExpressions.flatMap { expr =>
+      val aggregateExpressionMap = resultExpressions.flatMap { expr =>
         expr.collect {
-          case agg: AggregateExpression => agg
+          case agg: AggregateExpression => (agg.canonicalized, agg.deterministic) -> agg
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala]: I think non-deterministic functions should not be deduplicated, e.g. `select max(a + rand()), max(a + rand()) from ...` should still eveluate 2 aggregate funcitions.

my suggestion:
```
val aggregateExpressions = resultExpressions.flatMap { expr =>
  expr.collect {
    case agg: AggregateExpression => agg
  }
}
val aggregateExpressionMap = aggregateExpressions.filter(_.deterministic).map { agg =>
  agg.canonicalized -> agg
}.toMap
```

Comment (maryannxue) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala]: Thank you, @cloud-fan! You've made a very good point! We should address non-deterministic correctly here. I have confirmed, though, that all aggregate functions should be deterministic (and fortunately Spark already guarantees this in the analysis stage). So I think we are safe to use "agg.canonicalized" without considering the "deterministic" field here, and so as not to be confusing. I have also added a code comment addressing this point in my latest commit.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala [hunk 2]
```diff
@@ -205,14 +205,17 @@ object PhysicalAggregation {
     case logical.Aggregate(groupingExpressions, resultExpressions, child) =>
       // A single aggregate expression might appear multiple times in resultExpressions.
       // In order to avoid evaluating an individual aggregate function multiple times, we'll
-      // build a set of the distinct aggregate expressions and build a function which can
+      // build a map of the distinct aggregate expressions and build a function which can
       // be used to re-write expressions so that they reference the single copy of the
-      // aggregate function which actually gets computed.
-      val aggregateExpressions = resultExpressions.flatMap { expr =>
+      // aggregate function which actually gets computed. Note that aggregate expressions
+      // should always be deterministic, so we can use its canonicalized expression as its
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala]: can you point out where the code is that guarantee this? I took a look at `CheckAnalysis` and seems Spark allow Aggregate to have non-deterministic expressions.

**Comment by maryannxue** [comment]
@cloud-fan Please see CheckAnalysis.scala:170. It checks the input expression of each aggregate expression to make sure that they are not another aggregate function and are deterministic.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala (hunk 2 above)
Comment (rednaxelafx) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala]: @cloud-fan Agreed. e.g. `first()` in Spark SQL is marked as nondeterministic right now (although for the case of `first()` I'd actually believe we should make it deterministic instead, but that's for another story)

Comment (maryannxue) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala]: So we are talking about two types of "non-deterministic" here:
1. Across-query non-deterministic but in-query deterministic, which means the same expression can produce different results over the same input between different runs, but should always give the same result within the same run. sum/avg on floating point numbers could be an example. Shall we make sure that "select sum(f) - sum(f) from t" always return 0? and similarly for "first()" maybe, should "select first_value(c) = first_value(c) over ..." always return true?
It is important to define the behavior first, which will lead to opposite approaches on how to handle the "deterministic" field here.
2. Across-query and in-query non-deterministic, which I don't think is allowed anyway.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala]: good point, I think `first` is a nondeterministic function, SQL Server also document it: https://docs.microsoft.com/en-us/sql/relational-databases/user-defined-functions/deterministic-and-nondeterministic-functions (search for FIRST_VALUE).

But here, although `first` is nondeterministic, I think `select first(a) - first(a) from ...` should return 0, as its result should be consistent in-query.

CC @gatorsmile @hvanhovell

Comment (maryannxue) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala]: Then I think the solution we have right now (in my latest ci) is what we need. But we need to refine the code comment there to further illustrate cases like "first_value". And in general we may need better documentation regarding the behavior of FIRST_VALUE and similar functions. Agree?

Comment (gatorsmile) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala]: If we have the same non-deterministic functions in the same projection, they should return the same value.

**Comment by gatorsmile** [comment]
@maryannxue Could you update the comments? 

LGTM except the code comments.

**Comment by maryannxue** [comment]
Thank you for the comment, @gatorsmile! Code comment updated.

**Comment by gatorsmile** [comment]
LGTM

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala]: That's not true, see Postgres:
```
cloud=# select random(), random();
      random       |      random      
-------------------+------------------
 0.360924747306854 | 0.90011026058346
(1 row)

cloud=# select random() - random();
      ?column?       
---------------------
 -0.0260558221489191
(1 row)
```

Seems we need a way to distinguish FIRST_VALUE and RANDOM. Both of them are nondeterministic, but `FIRST_VALUE` has consistent values in the same project/aggregate.

Comment (gatorsmile) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala]: I did the check in Oracle. They return the same values.

**Comment by gatorsmile** [comment]
After I rethinking about it, non-determinstic expressions should not qualify this rule. Please remove it. Thanks!

**Comment by cloud-fan** [comment]
+1, before we figure out the semantic of nondeterministic aggregate functions, let's be careful here.

**Comment by maryannxue** [comment]
The point I was trying to make is, there are two types of non-deterministic aggregate functions: first being non-deterministic but not necessarily arbitrary, second being "deliberately" arbitrary, such as RAND. I think for the first category of non-deterministic functions, it's intuitive to expect that two occurrences of FIRST_VALUE(x) return the same value when user writes sth. like "SELECT FIRST(x), FIRST(x) + 1 FROM t".
I'd propose two options:
1. Take a step back and do not deduplicate non-deterministic functions as @cloud-fan first suggested.
1. To have a way to distinguish arbitrary functions (for UDF as well) for other non-deterministic cases, and avoid deduplication only for arbitrary functions.
Thoughts?

**Comment by gatorsmile** [comment]
Maybe we can first try the option 1 first. 

Did an offline discussion with @rednaxelafx . We can use `EquivalentExpressions`, which automatically treats `non-deterministic` expressions as different.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala [hunk 3]
```diff
@@ -235,8 +236,16 @@ object PhysicalAggregation {
         expr.transformDown {
           case ae: AggregateExpression =>
             // The final aggregation buffer's attributes will be `finalAggregationAttributes`,
-            // so replace each aggregate expression by its corresponding attribute in the set:
-            ae.resultAttribute
+            // so replace each aggregate expression by its corresponding attribute in the set.
+            // Note that non-deterministic aggregate expressions should not be deduplicated and
+            // should be handled differently.
+            val newAe = if (ae.deterministic) {
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala]: Do we still need to consider this? They will be different anyway, right?

Comment (gatorsmile) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala]: nvm. It does not contain it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala [hunk 4]
```diff
@@ -235,8 +236,16 @@ object PhysicalAggregation {
...
+              equivalentAggregateExpressions.getEquivalentExprs(ae)
+                .head.asInstanceOf[AggregateExpression]
+            } else {
+              ae
+            }
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala]: -> `equiExprs.getEquivalentExprs(ae).headOption.getOrElse(ae).asInstanceOf[AggregateExpression]`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala [hunk 5]
```diff
@@ -205,14 +205,15 @@ object PhysicalAggregation {
...
-      // be used to re-write expressions so that they reference the single copy of the
-      // aggregate function which actually gets computed.
+      // build a set of semantically distinct aggregate expressions and re-write expressions so
+      // that they reference the single copy of the aggregate function which actually gets computed.
+      val equivalentAggregateExpressions = new EquivalentExpressions
       val aggregateExpressions = resultExpressions.flatMap { expr =>
         expr.collect {
-          case agg: AggregateExpression => agg
+          case agg: AggregateExpression
+            if (!equivalentAggregateExpressions.addExpr(agg)) => agg
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala]: Add a comment above this line? `addExpr() returns false when the expression is non-deterministic`

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 6]
```diff
@@ -2677,4 +2678,29 @@ class SQLQuerySuite extends QueryTest with SharedSQLContext {
...
   }
+
+  test("SRARK-22266: the same aggregate function was calculated multiple times") {
+    val query = "SELECT a, max(b+1), max(b+1) + 1 FROM testData2 GROUP BY a"
+    val df = sql(query)
+    val physical = df.queryExecution.sparkPlan
+    val aggregates = physical.collect {
+      case agg : HashAggregateExec => agg
+    }
+    aggregates.foreach { agg =>
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: nit:
```
assert(aggregates.length == 1)
assert(aggregates.head.aggregateExpressions.size == 1)
```
If the implementation changed and we execute the query with other aggregate implementations, we should make sure this test is not ignored.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 7]
```diff
@@ -2677,4 +2678,29 @@ class SQLQuerySuite extends QueryTest with SharedSQLContext {
...
+  }
+
+  test("Non-deterministic aggregate functions should not be deduplicated") {
+    val query = "SELECT a, first_value(b), first_value(b) + 1 FROM testData2 GROUP BY a"
+    val df = sql(query)
+    val physical = df.queryExecution.sparkPlan
+    val aggregates = physical.collect {
+      case agg : HashAggregateExec => agg
+    }
+    aggregates.foreach { agg =>
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: ditto

**Comment by cloud-fan** [comment]
LGTM except one comment, thanks for working on it!

**Comment by gatorsmile** [comment]
retest this please

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**